habitat-baselines/habitat_baselines/rl/ppo/policy.py

```diff
@@ -83,6 +83,68 @@ def env_actions(self) -> torch.Tensor:
             return self.take_actions
 
 
+@dataclass
+class MultiAgentPolicyActionData(PolicyActionData):
+    """
+    Information returned from the `Policy.act` method representing the
+    information from multiple agent's action. This class is needed to store
+    actions of multiple agents together
+
+    :property should_inserts: Of shape [# envs, 1]. If False at environment
+        index `i`, then don't write this transition to the rollout buffer. If
+        `None`, then write all data.
+    :property policy_info`: Optional logging information about the policy per
+        environment. For example, you could log the policy entropy.
+    :property take_actions`: If specified, these actions will be executed in
+        the environment, but not stored in the storage buffer. This allows
+        exectuing and learning from different actions. If not specified, the
+        agent will execute `self.actions`.
+    :property values: The actor value predictions. None if the actor does not predict value.
+    :property actions: The actions to store in the storage buffer. if
+        `take_actions` is None, then this is also the action executed in the
+        environment.
+    :property rnn_hidden_states: Actor hidden states.
+    :property action_log_probs: The log probabilities of the actions under the
+        current policy.
+    """
+
+    rnn_hidden_states: torch.Tensor
```

Comment:
You shouldn't redefine all these attributes from the base class.

Reply:
> I thought an agent with a high level and a low level policy would be represented as two agents. It seems kind of hard to understand that we need to have MultiDiscrete action spaces while none of our actors support MultiDiscrete actions.

I agree. This is quite hacky. What is happening here is that the code is set so that we can reuse `ppo_trainer.py` even if we are working with multiple agents. The policy_action_space [here](https://github.com/facebookresearch/habitat-lab/blob/1fc7aa5229da2c882b73efdd057e948d72cdb355/habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py#L320) contains the union of action spaces among all the agents. Even if from the perspective of ppo_trainer this is only a single agent, internally we are updating two policies, working with two rollout buffers etc.

If the two agents have discrete action spaces, we need to use multidiscrete, where each categorical corresponds to one of the agents. If they use discrete + continuous, I am using TensorDict.